tutorials-v5/time-evolution/013_nonmarkovian_monte_carlo.md

```diff
@@ -0,0 +1,468 @@
+
+
+def Gamma(t):
+    return -2 * np.real(cd(t) / c(t))
+```
+
+Define the time interval. The initial time is shifted to avoid negative values of $\Gamma(t)$ at $t=0$ (which is unphysical):
+
+```python
+ti = optimize.root_scalar(Gamma, bracket=(1.4, 1.5)).root
+duration = 10
+steps = 100
+
+times = np.linspace(ti, ti + duration, steps + 1)
+```
+
+We plot the functions $\Gamma(t)$ and $S(t)$ over the time interval, demonstrating that $\Gamma(t)$ becomes negative:
+
+```python
+Gamma_values = np.array([Gamma(t) for t in times])
+S_values = np.array([S(t) for t in times])
+
+plt.plot(times - ti, Gamma_values, label=r"$\Gamma(t)$")
+plt.plot(times - ti, S_values, label=r"$S(t)$")
+plt.xlabel(r"$t$")
+plt.legend()
+plt.show()
+```
+
+To speed up the following calculation by a factor 3-4, we store $\Gamma(t)$ and $S(t)$ as interpolations.
+
+```python
+Gamma_tmp = CubicSpline(times, Gamma_values)
+S_tmp = CubicSpline(times, S_values)
+
+
+def Gamma_int(t):
+    return float(Gamma_tmp(t))
+
+
+def S_int(t):
+    return float(S_tmp(t))
+```
+
+##### Monte-Carlo Simulation
+
+We specify some numerical parameters, i.e., the number of trajectories and whether to use parallel computing. Then we define the Hamiltonian and the jump operator corresponding to the master equation, and the initial state.
+
+```python
+options = {"map": "parallel", "progress_bar": "enhanced"}
+ntraj = 2500
+
+H = 2 * qt.sigmap() * qt.sigmam()
+ops_and_rates = [[qt.sigmam(), Gamma_int]]
+psi0 = qt.basis(2, 0)
+e_ops = [H]
+```
+
+In analogy with `mcsolve` and the corresponding `MCSolver` class, we may either instantiate a `NonMarkovianMCSolver` object and call its `run` or `start` / `step` methods to run the Monte-Carlo simulation, or make use of the `nm_mcsolve` convenience function.
+Here, we construct a `NonMarkovianMCSolver` in order to explicitly verify that the solver ensures that the jump operators satisfy the completeness relation $\sum_n A^\dagger_n A_n \propto 1$.
+We can see that a second jump operator is added in order to satisfy this relation.
+
+```python
+solver = qt.NonMarkovianMCSolver(qt.QobjEvo([H, S_int]),
+                                 ops_and_rates,
+                                 options=options)
+
+for L, _ in solver.ops_and_rates:
+    print(L, "\n")
+
+completeness_check = sum(L.dag() * L for L, _ in solver.ops_and_rates)
+with qt.CoreOptions(atol=1e-5):
+    assert completeness_check == qt.qeye(2)
+```
+
+Finally, we run the Monte-Carlo simulation:
+
+```python
+MCSol = solver.run(psi0, tlist=times, ntraj=ntraj, e_ops=e_ops)
+```
+
+##### Benchmark against non-stochastic simulation
+
+We will compare our results to the following exact `mesolve` simulation:
+
+```python
+d_ops = [[qt.lindblad_dissipator(qt.sigmam(), qt.sigmam()), Gamma]]
+MESol = qt.mesolve([H, S], psi0, times, d_ops, e_ops,
+                   options={key: val for key, val in options.items()
+                            if key in qt.MESolver.solver_options})
```

Comment:
```suggestion
                   options={key: val for key, val in options.items()
                            if key in qt.MESolver.solver_options})
```
We could avoid this incantation by having `options` which has the common options and `nm_options` which has the extra options for `nm_mcsolve`.